### PR TITLE
feat: export callbacks for hooks

### PR DESCRIPTION
**Pull request by ph-fritsche**
### What

Export `@experimental` functions `prepare`, `reset` and `detach` to improve support for different test environments.

```js
import userEvent, {prepare, reset, detach } from '@testing-library/user-event'
import { beforeAll, afterEach, afterAll, test } from 'some-test-runner'
import { DomImplementation } from 'some-dom-implementation'

const window = new DomImplementation()

beforeAll(() => prepare(window))
afterEach(() => reset(window))
afterAll(() => detach(window))

test('something', async () => {
  const user = userEvent.setup({document: window.document})
  // ...
```

### Why

UserEvent should work in various environments. Exposing these functions allows our users to integrate UserEvent in environments that don't use global hooks or global document.

### How

Includes a fix to `restoreFocus`.

Includes a check if `globalThis.window` exists.

### Checklist
- [ ] Ready to be merged

### CODE REVIEW

**Comment by brynshanahan** [comment]
How about calling `userEvent.setup()` and `userEvent.${directApis}` calls the prepare functions automatically then, you would only need to `afterEach(() => detach())`

**Comment by ph-fritsche** [comment]
@brynshanahan `prepare` is already called internally by `setup`. Exposing the function allows to call it earlier to resolve conflicts with other frameworks/tools.
```
This should work:
domImplementation.createDocument()
--> userEvent.prepare()
--> someLibrary.setup()/userEvent.setup()
  --> someComponent.render()
  --> test interaction
  --> someLibrary.teardown()/userEvent.reset() (reverse order)
[...]
--> userEvent.detach()
--> execute any code on the document as if userEvent had never been used
```
